Sources/Plasma/NucleusLib/pnAsyncCoreExe/pnAceDns.cpp

```diff
@@ -0,0 +1,155 @@
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Additional permissions under GNU GPL version 3 section 7
+
+If you modify this Program, or any covered work, by linking or
+combining it with any of RAD Game Tools Bink SDK, Autodesk 3ds Max SDK,
+NVIDIA PhysX SDK, Microsoft DirectX SDK, OpenSSL library, Independent
+JPEG Group JPEG library, Microsoft Windows Media SDK, or Apple QuickTime SDK
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#include "Pch.h"
+
+using tcp = asio::ip::tcp;
+
+struct DnsResolver
+{
+    asio::io_context m_context;
+    asio::executor_work_guard<asio::io_context::executor_type> m_workGuard;
+    tcp::resolver m_resolver;
+    std::thread m_lookupThread;
+
+    DnsResolver()
+        : m_context(), m_resolver(m_context), m_workGuard(m_context.get_executor())
+    {
+        // Start the resolver thread
+        m_lookupThread = std::thread([this] { m_context.run(); });
+    }
+
+    void destroy(unsigned exitThreadWaitMs)
+    {
+        m_workGuard.reset();
+        m_resolver.cancel();
+        AsyncThreadTimedJoin(m_lookupThread, exitThreadWaitMs);
+        m_lookupThread = {};
+    }
+};
+
+struct DnsResolveData
+{
+    ST::string          m_name;
+    FAsyncLookupProc    m_lookupProc;
+    void*               m_param;
+};
+
+static std::recursive_mutex     s_critsect;
+static DnsResolver             *s_resolver = nullptr;
+
+void DnsDestroy(unsigned exitThreadWaitMs)
+{
+    if (s_resolver) {
+        s_resolver->destroy(exitThreadWaitMs);
+        delete s_resolver;
+        s_resolver = nullptr;
+    }
+}
+
+static void AddressResolved(const asio::error_code& err,
+                            const tcp::resolver::results_type& results,
+                            const DnsResolveData& data)
+{
+    if (err || results.empty()) {
+        if (err)
+            LogMsg(kLogFatal, "DNS: Failed to resolve {}: {}", data.m_name, err.message());

```

Comment:
Can we move towards using `plNetApp::StaticErrorMessage` instead of pnAsyncCore's `LogMsg`? I'd like to be able to get rid of `LogMsg` entirely to break the pnAsyncCore dependency on plStatusLog

Reply:
One thing to consider with these log messages is that any `plStatusLog`-backed logger (including `plNetApp::Static*Message` is that they block. Originally, `LogMsg` did not block, so we'll want to be mindful of that.

Reply:
Alas, linking in plNetApp also requires some creatable registration which is unavailable in plUruLauncher...  So this cannot be done yet.

Reply:
IMO, it would be better to demote plStatusLog to pnStatusLog if we're going to be using it down here in NucleusLand.